stan/math/prim/mat/functor/mpi_parallel_call.hpp

```diff
@@ -0,0 +1,488 @@
+ *    are locally cached such that static data is only transferred on
+ *    the first evaluation. Note that the work is equally distributed
+ *    among the workers. That is N jobs are distributed ot a cluster
+ *    of size W in N/W chunks (the remainder is allocated to node 1
+ *    onwards which ensures that the root node 0 has one job less).
+ * 5. Once the parameters and static data is distributed, the reduce
+ *    operation is applied per defined job. Each job is allowed to
+ *    return a different number of outputs such that the resulting
+ *    data structure is a ragged array. The ragged array structure
+ *    becomes known to mpi_parallel_call during the first evaluation
+ *    and must not change for future calls.
+ * 6. Finally the local results are gathered on the root node over MPI
+ *    and given on the root node to the combine functor along
+ *    with the ragged array data structure.
+ *
+ * The MPI cluster resource is aquired with construction of
+ * mpi_parallel_call and is freed once the mpi_parallel_call goes out
+ * of scope (that is, deconstructed).
+ *
+ * Note 1: During MPI operation everything must run synchronous. That
+ * is, if a job fails on any of the workers then the execution must
+ * still continue on all other workers. In order to maintain a
+ * synchronized state, even the worker with a failed job must return a
+ * valid output chunk since the gather commands issued on the root to
+ * collect results would otherwise fail. Thus, if a job fails on any
+ * worker the a respective status flag will be transferred after the
+ * reduce and the results gather step.
+ *
+ * Note 2: During the first evaluation of the function the ragged
+ * array sizes need to be collected on the root from all workers. This
+ * is needed on the root such that the root knows how many outputs are
+ * computed on each worker. This information is then cached for all
+ * subsequent evaluations. However, caching this information can only
+ * occur if and only if the evaluation of all functions was
+ * successfull on all workers during the first run. Thus, the first
+ * evaluation is handled with special care to ensure that caching of
+ * this meta info is only done when all workers have successfully
+ * evaluated the function and otherwise an exception is raised.
+ *
+ * @tparam call_id label for the static data
+ * @tparam ReduceF reduce function called for each job
+ * @tparam CombineF combine function called on the combined results on
+ * each job along with the ragged data structure information
+ */
+template <int call_id, typename ReduceF, typename CombineF>
+class mpi_parallel_call {
+  boost::mpi::communicator world_;
+  const std::size_t rank_ = world_.rank();
+  const std::size_t world_size_ = world_.size();
+  std::unique_lock<std::mutex> cluster_lock_;
+
+  // local caches which hold local slices of data
+  typedef internal::mpi_parallel_call_cache<call_id, 1,
```

Comment:
I agree with the `_t` convention, but here I deliberately deviate from that since the type definition is the same as declaring data. This is odd, I know, but the typedef is a declaration of a data item which exists a single time in the program. That is why I would prefer to stay without the `_t` here.